File: cartmodel.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from datamodel import DataModel
from foodmodel import FoodModel


class CartModel(DataModel):
	def __init__(self, id = None):
		self.prefix = 'cartmodel_obj_string_'
		DataModel.__init__(self, id)
		self.food_ids = dict()
		self.food_count = 0
		self.total = 0
		self.is_locked = False
		self.is_bad_order = False

	def load(self):
		if DataModel.load(self) is False: return False
		data_dict = self.data_dict
		tmp = zip(data_dict['food_ids'].split(','), data_dict['food_nums'].split(','))
		self.food_ids = {x: int(y) for x, y in tmp if y != ''}
		self.food_count = int(data_dict['food_count'])
		self.total = int(data_dict['total'])
		self.is_locked = (data_dict['is_locked'] == "1")
		self.is_bad_order = (data_dict['is_bad_order'] == "1")
		self.userid = data_dict['userid']

	def save(self):
		data_dict = self.data_dict
		data_dict['food_count'] = self.food_count
		data_dict['total'] = self.total
		data_dict['is_locked'] = "1" if self.is_locked else "0"
		data_dict['is_bad_order'] = "1" if self.is_bad_order else "0"
		data_dict['userid'] = self.userid
		data_dict['food_ids'] = ','.join(map(lambda i: str(i), self.food_ids.keys()))
		data_dict['food_nums'] = ','.join(map(lambda i: str(i), self.food_ids.values()))
		DataModel.save(self)

	def add_food(self, id, count):
		if self.is_locked:
			return '''{
				"code": "ORDER_LOCKED",
				"message": "订单已经提交不能修改"
			}'''
		if self.food_count + count > 3:
			return '''{
				"code": "FOOD_OUT_OF_LIMIT",
				"message": "篮子中食物数量超过了三个"
			}'''
		# TODO: lock
		food = FoodModel.fetch(id)
		if food.reserve(count):
			self.food_count += count
			self.total += count * food.price
			try:
				self.food_ids[id] += count
			except KeyError:
				self.food_ids[id] = count
		else:
			self.is_bad_order = True
		self.save()
		return True

File: src/eleme/cart_model.go
package eleme

import (
	"fmt"
	"strings"
	"encoding/json"
	"github.com/garyburd/redigo/redis"
)

type CartModel struct {
	Id string
	Userid string
	FoodIds map[string] int
	FoodCount int
	Total int
}

func (this *CartModel) save(redisConn *redis.Conn) {
	raw, _ := json.Marshal(this)
	(*redisConn).Do("SET", this.Id, string(raw))
}

func (this *CartModel) load(redisConn *redis.Conn) bool {
	raw, _ := redis.String((*redisConn).Do("GET", this.Id))
	if raw == "" { return false }
	json.Unmarshal([]byte(raw), this)
	return true
}

func createCart(userid, cartid string) *CartModel {
	return &CartModel{cartid, userid, map[string]int{}, 0, 0}
}

func (this *CartModel) fetch(redisConn *redis.Conn, cartid string) *CartModel {
	ret := new(CartModel)
	ret.Id = cartid
	if ret.load(redisConn) == false { return nil }
	return ret
}

func addFood(redisConn *redis.Conn, cartid, userid string, food *FoodModel, count int) string {
	ret, _ := redis.String((*redisConn).Do("EVALSHA", add_food_script_sha1, 5, cartid, "food_stock_of_" + food.id, "empty_cart", cartid + "_is_bad_order", "userid2orderid_" + userid, userid, food.realidString, count, food.price))
	return ret
}

func userid2orderid(redisConn *redis.Conn, userid string) string {
	ret, _ := redis.String((*redisConn).Do("GET", "userid2orderid_" + userid))
	return ret
}

func makeOrder(redisConn *redis.Conn, cartid, userid string) string {
	ret, _ := redis.String((*redisConn).Do("EVALSHA", make_order_script_sha1, 2, cartid + "_is_bad_order", "userid2orderid_" + userid, cartid))
	switch ret {
		case "1":
			return "{\"code\":\"FOOD_OUT_OF_STOCK\",\"message\":\"食物库存不足\"}"
		case "2":
			return "{\"code\":\"ORDER_OUT_OF_LIMIT\",\"message\":\"每个用户只能下一单\"}"
		default:
			return ""
	}
}

func (this *CartModel) dump() string {
	var buf []string
	for id, count := range this.FoodIds {
		buf = append(buf, fmt.Sprintf("{\"food_id\":%s,\"count\":%d}", id, count))
	}
	ret := "{\"id\":\"" + this.Id + "\",\"user_id\":" + this.Userid[5:] + ","
	ret += "\"items\":[" + strings.Join(buf, ",") + "],"
	ret += fmt.Sprintf("\"total\":%d}", this.Total)
	return ret
}

func (this *CartModel) dumpAll(redisConn *redis.Conn) string {
	var buf []string
	for _, user := range username2user {
		orderId := userid2orderid(redisConn, user.id)
		if orderId != "" {
			cart := cartModel.fetch(redisConn, orderId)
			if cart != nil {
				buf = append(buf, cart.dump())
			} else {
				buf = append(buf, "{\"id\":\"" + this.Id + "\",\"user_id\":" + this.Userid[5:] + ",\"items\":[],\"total\":0}")
			}
		}
	}
	return "[" + strings.Join(buf, ",") + "]"
}

var cartModel *CartModel
